### ISSUE TITLE
Unable to load libvips-42.dll

### ISSUE THREAD

**Issue by StefanTargos** [libvips/pyvips#194]
Hello,

I recently installed pyvips in accordance to the description in https://pypi.org/project/pyvips/. However, when I try to import pyvips, I always get the very same error.

**OSError: cannot load library 'C:\MorePythonSidePackages\vips-dev-8.9\bin\libvips-42.dll': error 0x7f**

I downloaded the windows binary vips-dev-w64-all-8.9.2.zip from https://github.com/libvips/libvips/releases, set the bin folder to my path and executed to command "pip install pyvips".

It is possible to run vips.exe via cmd but not via python.

I am using windows 10 and python3.6.7.

I also tried different versions of vips but i always get the very same error.

The above mentioned file 'C:\MorePythonSidePackages\vips-dev-8.9\bin\libvips-42.dll' does exit, it can just not be loaded. 

The complete error message is the following:

Traceback (most recent call last):
  File "C:\Program Files\Python36\lib\site-packages\pyvips\__init__.py", line 19, in <module>
    import _libvips
ModuleNotFoundError: No module named '_libvips'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:/Targos_AI/pathology-streaming-pipeline/streaming/train.py", line 21, in <module>
    from streaming.tissue_dataset import TissueDataset
  File "C:\Targos_AI\pathology-streaming-pipeline\streaming\tissue_dataset.py", line 8, in <module>
    import pyvips
  File "C:\Program Files\Python36\lib\site-packages\pyvips\__init__.py", line 71, in <module>
    vips_lib = ffi.dlopen(_vips_libname)
  File "C:\Program Files\Python36\lib\site-packages\cffi\api.py", line 150, in dlopen
    lib, function_cache = _make_ffi_library(self, name, flags)
  File "C:\Program Files\Python36\lib\site-packages\cffi\api.py", line 832, in _make_ffi_library
    backendlib = _load_backend_lib(backend, libname, flags)
  File "C:\Program Files\Python36\lib\site-packages\cffi\api.py", line 828, in _load_backend_lib
    return backend.load_library(path, flags)
OSError: cannot load library 'C:\MorePythonSidePackages\vips-dev-8.9\bin\libvips-42.dll': error 0x7f

Process finished with exit code 1

When I try to load a different dll file using the command "backend.load_library(path, flags)" I get no error.

I searched previouse Issues, such as #44 but they don't solve my problem. 
Do you have any suggestions what I could try?

Thank you in advance

**Comment by jcupitt** [comment]
Hello @StefanTargos,

libvips.dll is 64-bit, so you need a 64-bit Python. Could that be the error?

**Comment by StefanTargos** [comment]
Hello John,
thank you for the very fast response. 
I am using Python 3.6.7 (v3.6.7:6ec5cf24b7, Oct 20 2018, 13:35:33) [MSC v.1900 64 bit (AMD64)] on win32, which It is 64-bit.

**Comment by jcupitt** [comment]
It seems `error 0x7f` means a function cannot be located.

Since you're not fetching a function, just opening the dll, it means that `libvips-42.dll` must be trying to load in other dlls as it starts up, and one of those dlls is missing a required function. That means it must be trying to link in the wrong version of one of the dlls.

I think you must have some other package installed which also has (guess!) `libglib-2.dll`, and that package is using an older version of glib which is missing some function libvips needs.

I would try removing libvips from your PATH, and removing anything else non-standard on there, and instead setting PATH from your Python program.

Put this near the start:

```python
import os
vipshome = 'C:/MorePythonSidePackages/vips-dev-8.9/bin'
os.environ['PATH'] = vipshome + ';' + os.environ['PATH']
import pyvips
```

It should reduce the chances of accidental dll clash. Hopefully!

**Comment by StefanTargos** [comment]
I already tried to add VIS to my python PATH using os.environ and removed it from my windows environment variables.
However, it is a good idea to remove different, non-standard paths from my windows environment variables too, as they may interfere with the dlls of VIS.

I will give it a try and post the results after that.
Right now I am not on my working machine. I will try your suggestion tomorrow morning.

**Comment by StefanTargos** [comment]
Your idea most solved my problem. Thank you so much for that. 
The problem was caused by the following packages:
Among others, I added openlide and ASAP (Automated Slide Analysis Platform) to my windows / python environment. 

In the following I will state how I installed these packages, how they caused problems and how the problems could be solved.

First I started with openslide. Simply installing it via pip istall was not sufficient to make it work. I always got the error: OSError: [WinError 126] The specified module could not be found.
I added additional dll files to my python path as described here: https://github.com/openslide/openslide-winbuild/issues/18

Next, I installed ASAP from here: https://github.com/computationalpathologygroup/ASAP/releases/tag/1.9

I added the binary of ASAP to my path and tried "import multiresolutionimageinterface".
This was not working, because there is one dll of ASAP which has the same name as one dll of openslide (which was copied to the python path).
So I created a work around and removed ASAP from my path but included the path in my python script.
os.environ["PATH"] = 'C:\\Program Files\\ASAP 1.9\\bin;' + os.environ["PATH"]
import multiresolutionimageinterface as mir # this is part of ASAP 1.9

Next, I wanted to install VIPS. I turns out, that there is another dll which has the same name as one dll of open slide.

To fix this, I had to remove the dlls of openslide from the python folder. I created a new folder, which contains the openslide dlls. This folder will be added to the path via an os.environ["PATH"] statement.

In short, non of the three packages must be added to the global windows environment path.

Scripts, which use openslide will begin with the statement:
import os
openSlideDlls = r'C:\Program Files\Python36\opslide_dlls'
os.environ['PATH'] = openSlideDlls + ';' + os.environ['PATH']

Scripts, which use ASAP will begin with the statement:
import os
os.environ["PATH"] = 'C:\\Program Files\\ASAP 1.9\\bin;' + os.environ["PATH"]
import multiresolutionimageinterface as mir # this is part of ASAP 1.9

And Scripts, which use VIPS (pyvips) will begin with:
import os
vipshome = r'C:\MorePythonSidePackages\vips-dev-8.9\bin'
os.environ['PATH'] = vipshome + ';' + os.environ['PATH']


Thank you again for your help. I will close this issue

**Comment by jcupitt** [comment]
Yes, adding DLLs to the python bin folder is likely to break other packages. It's much better to modify PATH.

libvips includes openslide, so you might not need both installed.
